### PR TITLE
Solaris/SmartOS platform support

### PR DESCRIPTION
**Pull request by doublerebel**
These commits are a full port of ArangoDB to Solaris/SmartOS.  Since Joyent supports both NodeJS and SmartOS, I believe Solaris/SmartOS is a sustainable target for ArangoDB.

Thanks to Arango's structure being mostly platform-agnostic with many helpful #define flags, this version is very close to the original.  I left the original logic and attempted to make changes only affecting Solaris, where possible.

I am most unsure about commit b68a430, **CODE, TRI_facilitynames not declared in codebase (?)**.  I cannot find where `CODE`, `_code`, or `TRI_facilitynames` would be defined in stdlib or the arangodb codebase.

I have run the test suite on this build, and it seems to pass with flying colors.

Thanks to all for your work!

### Build Instructions on a new SmartOS instance
```sh
# sudo su
pkgin -y upgrade && pkgin -y install git build-essential gcc49 go
rm /opt/local/bin/gcc && ln -s /opt/local/gcc49/bin/gcc /opt/local/bin/gcc
rm /opt/local/bin/g++ && ln -s /opt/local/gcc49/bin/g++ /opt/local/bin/g++
# exit

git clone --single-branch --depth 1 git://github.com/doublerebel/arangodb.git
cd arangodb
make setup
./configure
LD_ALTEXEC="/home/admin/arangodb/utils/gld-wrapper" make
```
Building on a "hi-cpu" 4 CPU/4 GB instance with -j 8 is reasonably quick.

### CODE REVIEW

File: lib/Basics/FileUtils.cpp [hunk 1]
```diff
@@ -291,8 +291,6 @@ bool copyDirectoryRecursive(std::string const& source,
                             std::string const& target, std::string& error) {
   bool rc = true;
 #ifdef TRI_HAVE_WIN32_LIST_FILES
-  auto isSymlink = [](struct _finddata_t item) -> bool { return false; };
```
Comment (dothebart) [inline lib/Basics/FileUtils.cpp]: removing this breaks windows support?

**Comment by dothebart** [comment]
is there a smart way available to get a jenkins build slave or travis with this?

**Comment by dothebart** [comment]
another tiny heads up, we're heading to ditch autofoo in favour of the cmake (which we already have had for a while) so that should work too.

**Comment by doublerebel** [comment]
It looks simple to setup a Joyent instance as a Jenkins slave, [Joyent uses it to build their own SmartDataCenter](https://github.com/joyent/mountain-gorilla/blob/master/tools/jenkins/sdc-jenkins/create-jenkins-slave/setup-jenkins-slave.sh).

There are [small free instances provided by Joyent for dev](https://www.joyent.com/portal/701800000015N22-toc.html).  The build would be slow but I could provision and donate one of mine as a build slave.

**Comment by doublerebel** [comment]
Thanks, I'll port these changes to cmake and add them to the PR.

**Comment by dothebart** [comment]
Just installed a VM with smartOS. Merged since there were conflicts; Will fix that syslog stuff.

**Comment by matthiasg** [comment]
just tried to compile the origin/3.0 branch on SmartOS "20151210T194528Z". Its not working anymore.

the new cmake process generates the make file fine, but make emits errors when compiling `libev` of all things:

```
In file included from /root/arangodb/3rdParty/libev/libev-4.22/ev.c:423:0:
/usr/include/sys/statfs.h:75:5: note: declared here
 int statfs(const char *, struct statfs *, int, int);
     ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4237:23: error: 'struct statfs' has no member named 'f_type'
                && (sfs.f_type == 0x1373 /* devfs */
                       ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4238:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x4006 /* fat */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4239:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x4d44 /* msdos */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4240:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0xEF53 /* ext2/3 */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4241:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x72b6 /* jffs2 */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4242:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x858458f6 /* ramfs */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4243:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x5346544e /* ntfs */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4244:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x3153464a /* jfs */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4245:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x9123683e /* btrfs */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4246:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x52654973 /* reiser3 */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4247:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x01021994 /* tmpfs */
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4248:26: error: 'struct statfs' has no member named 'f_type'
                    || sfs.f_type == 0x58465342 /* xfs */))
                          ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c: In function 'infy_cb':
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4216:66: error: 'NAME_MAX' undeclared (first use in this function)
 # define EV_INOTIFY_BUFSIZE (sizeof (struct inotify_event) * 2 + NAME_MAX)
                                                                  ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4343:13: note: in expansion of macro 'EV_INOTIFY_BUFSIZE'
   char buf [EV_INOTIFY_BUFSIZE];
             ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4216:66: note: each undeclared identifier is reported only once for each function it appears in
 # define EV_INOTIFY_BUFSIZE (sizeof (struct inotify_event) * 2 + NAME_MAX)
                                                                  ^
/root/arangodb/3rdParty/libev/libev-4.22/ev.c:4343:13: note: in expansion of macro 'EV_INOTIFY_BUFSIZE'
   char buf [EV_INOTIFY_BUFSIZE];
             ^
3rdParty/libev/CMakeFiles/ev.dir/build.make:62: recipe for target '3rdParty/libev/CMakeFiles/ev.dir/libev-4.22/ev.c.o' failed
make[2]: *** [3rdParty/libev/CMakeFiles/ev.dir/libev-4.22/ev.c.o] Error 1
CMakeFiles/Makefile2:150: recipe for target '3rdParty/libev/CMakeFiles/ev.dir/all' failed
make[1]: *** [3rdParty/libev/CMakeFiles/ev.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by matthiasg** [comment]
I was able to fix `libev` by patching ev.c. It turns out it was due to cmake detecting inotify, a small fix to ev.c was able to overcome this issue.

It was already reported on in the libev group here. http://lists.schmorp.de/pipermail/libev/2016q1/002641.html

**Comment by matthiasg** [comment]
There also is an issue with the make file generated for v8.

```
#  make VERBOSE=1
...
cd /root/arangodb/3rdParty/V8/V8-5.0.71.39 && make -C /root/arangodb/build/3rdParty/V8/v8 -f Makefile.x64.release BUILDTYPE=Release builddir=/root/arangodb/build/3rdParty/V8/v8/x64.release CC=/opt/local/bin/cc CXX=/opt/local/bin/c++ "CFLAGS= -fno-delete-null-pointer-checks" "CXXFLAGS= -fno-delete-null-pointer-checks" LDFLAGS= PYTHON_EXECUTABLE=/opt/local/bin/python icui18n icuuc icudata v8
...
   from ../../../../3rdParty/V8/V8-5.0.71.39/third_party/icu/source/common/umapfile.c:22:
/usr/include/sys/feature_tests.h:391:2: error: #error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications       require the use of c99"
#error "Compiler or options invalid; UNIX 03 and POSIX.1-2001 applications \
```

i tried patching to use gcc and also adding -std=c99 to CFLAGS but it didnt work out yet

v8 and libev are both available as compiled packages maybe there is a way to tell arangodb to use the compiled binaries instead of trying to compile itself alternatively.

**Comment by dothebart** [comment]
@matthiasg some parts may have been lost in the V8 upgrade. Did you fix the gyp files inside of V8? Since at least some of them have special treatments for Solaris11 (which we added as a followup of this PR) ?

**Comment by matthiasg** [comment]
@dothebart no i did not have time to look through the entire makefile chain since this was a proof project only. I just wrapped all gcc calls using a script to ensure -std=c99 was added.

**Pull request merged**